canard.cpp

```diff
@@ -0,0 +1,186 @@
+#include <libcanard/canard.h>
+#include <cstdint>
+#include <libcanard/drivers/stm32/canard_stm32.h>
+#include <libcanard/helpers/RingBuffer.h>
+#include <libcanard/helpers/helpers.h>
+
+static RingBuffer<CanardCANFrame, CAN_RX_RB_SIZE> CAN_RX_RB;
+static RingBuffer<CanardCANFrame, CAN_TX_RB_SIZE> CAN_TX_RB;
+
+int getRXSize() {
+    return CAN_RX_RB.size();
+}
+
+int getTXSize() {
+    return CAN_TX_RB.size();
+}
+
+
+extern "C" {
+extern CanardInstance canard;
+
+volatile int allTimeMaxRxSize;
+volatile int allTimeMaxTxSize;
+volatile int maxRxSize;
+volatile int maxTxSize;
+
+#if defined(STM32F0) || defined(STM32F3)
+void USED CAN_IT_Callback() {
+    if (!CAN_RX_RB.is_full()) {
+        canardSTM32Receive(&CAN_RX_RB.get_next_write_entry());
+        CAN_RX_RB.advance_write();
+    } else {
+        canardSTM32ReleaseFIFO();
+        canard_errors.rx_errors++;
+    }
+
+    volatile int rxSize = getRXSize();
+    if (rxSize > maxRxSize) {
+        maxRxSize = rxSize;
+    }
+    if (rxSize > allTimeMaxRxSize) {
+        allTimeMaxRxSize = rxSize;
+    }
+    canardSTM32Receive_IT();
+}
+#endif
+
+volatile bool canTXTransmitting = false;
+
+volatile canard_errors_t canard_errors = {
+    .tx_errors = 0,
+    .rx_errors = 0,
+    .broadcast_errors = 0
+};
+
+void processRxQueue() {
+    const uint64_t ts_usec = getMonotonicTimestampMSec() * 1000UL;
+
+    while (!CAN_RX_RB.is_empty()) {
+        canardHandleRxFrame(&canard, &CAN_RX_RB.get_next_read_entry(), ts_usec);
+        CAN_RX_RB.advance_read();
+    }
+}
+
+void processTxQueue(void) {
```

Comment:
These definitions were moved from mordor, so their declarations should come along with them (they still exist in firmwares/lib/formlabs/inc/CAN_/CAN_c.h, which is confusing).

We also ought to prefix these with "canard" like the other exposed functions in canard.c.